Schema enums follow up (part 2)

austindrenski opened the pull request:
Follow-up to #605 and #626.

YohDeadfall commented on test/EFCore.PG.Tests/Storage/NpgsqlTypeMappingTest.cs:
```diff
@@ -333,14 +333,28 @@ public void GenerateSqlLiteral_returns_json_literal()
+                    null,
+                    typeof(DummyEnum),
+                    new NpgsqlSqlGenerationHelper(new RelationalSqlGenerationHelperDependencies()),
+                    new NpgsqlSnakeCaseNameTranslator());
```
Move `new` one line up and decrease indentation for the following lines.

YohDeadfall commented on test/EFCore.PG.Tests/Storage/NpgsqlTypeMappingTest.cs:
```diff
@@ -333,14 +333,28 @@ public void GenerateSqlLiteral_returns_json_literal()
+                    null,
+                    typeof(DummyEnum),
+                    new NpgsqlSqlGenerationHelper(new RelationalSqlGenerationHelperDependencies()),
+                    new NpgsqlSnakeCaseNameTranslator());
```
And here too.

YohDeadfall commented on src/EFCore.PG/Metadata/PostgresEnum.cs:
```diff
@@ -11,10 +11,10 @@ namespace Npgsql.EntityFrameworkCore.PostgreSQL.Metadata
-        readonly IAnnotatable _annotatable;
-        readonly string _annotationName;
+        [NotNull] readonly IAnnotatable _annotatable;
+        [NotNull] readonly string _annotationName;
```
I hope these attributes will go away very soon.

austindrenski commented:
@roji You asked [previously](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/626#discussion_r234516176) about some code that looked like this:

```c#
// Alter operations may not have a default schema attached. If not, try to attach one for schema-qualified types.
if (operation[RelationalAnnotationNames.DefaultSchema] == null)
    operation[RelationalAnnotationNames.DefaultSchema] = model[RelationalAnnotationNames.DefaultSchema];
```

I think my previous answer was incomplete. Completely removing that code in #626 worked because #605 had some [(unintentionally) hackish code](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/605/files#diff-7d6fd89c6ae4f223fc36e554f5663a2dR723) that covered up the root of the issue by accessing the model as a fallback when a schema was unavailable for an enum.

It was a bit confusing to unwind, but I think the root of the issue is that we weren't yielding the default schema annotation from `NpgsqlMigrationsAnnotationProvider.For(...)` which meant that `AlterDatabaseOperation` wouldn't have access to it, and so a `PostgresEnum` or `PostgresRange` on that operation wouldn't have access to it either. 

I've changed that here, such that `NpgsqlMigrationsAnnotationProvider.For(...)` now yields the default schema annotation, and everything is back to passing (without the hacks).

austindrenski commented on src/EFCore.PG/Migrations/Internal/NpgsqlMigrationsAnnotationProvider.cs:
```diff
@@ -55,6 +55,7 @@ public override IEnumerable<IAnnotation> For(IIndex index)
 
         public override IEnumerable<IAnnotation> For(IModel model)
             => model.GetAnnotations().Where(a =>
+                a.Name.StartsWith(RelationalAnnotationNames.DefaultSchema) ||
```
I've looked into this a few times, and I could be missing something....but it just looks like their provider doesn't have to handle user-defined types that need to be stored as annotations.

In some ways, these annotations are quasi-dependent on each other, which feels like a problem that _ought_ to be encountered by other providers. But apparently not so far?

roji replied:
Here's what's bothering me, and it could impact the general approach of this PR...

Including a DefaultSchema annotation on a *migration operation* (which is what it means for to be in `NpgsqlMigrationsAnnotationProvider`) implies that there has been a change in the default schema, and which needs to be somehow applied to the database via SQL. For example, if PostgreSQL had some of `SET default_schema` parameter which persists in the database, we would use an AlterModel migration operation to change that, and it would make sense to flow the DefaultSchema annotation on that operation.

Moreover, looking at the latest state of this PR, the DefaultSchema annotation doesn't seem to be used by NpgsqlMigrationsSqlGenerator... `GenerateEnumStatements()` iterates over all enum annotations on the AlterDatabaseOperation, which constructs PostgresEnum instances with fully-provided annotation name (already containing the schema) and value as strings. In other words, it seems that `BuildAnnotationName()` doesn't actually get called when generating migration SQL. So unless I'm missing something this can be removed?

One more concern I've stumbled upon when looking into this... When constructing the annotation, `BuildAnnotationName()` builds the annotation name with the schema - it tries to see if a schema was explicitly provided to the enum, defaults to the DefaultSchema annotation, and finally to null. So in the regular case (no schema, no DefaultSchema), we generate an annotation name such as `Npgsql:Enum:MyEnum`, no schema. Now, when the MigrationsSqlGenerator picks this annotation up, it calls the `Schema` property to get the enum's schema, but at that point you consult the DefaultSchema annotation again. This seem incorrect - if the enum annotation was *created* without a schema (i.e. the annotation name contains no schema), then it seems like we shouldn't override that when retrieving the schema later.

In other words, unless I'm missing something, I think it's safer and more correct to consult the DefaultSchema only at creation time. If this is right, we need to fix this in range as well.

austindrenski replied:
@roji Thanks for laying this all out. It was very helpful feedback. After some more thought, I agree that something isn't quite right with the current approach.

I think it stems from an early decision to have the `DefaultSchema` annotation eagerly bound into the enum annotation. That decision created the dependency between those annotations, the need for `DefaultSchema` to be passed to the migration operation, __and__ the inconsistencies in the annotation keys that I [mentioned](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/626#issuecomment-434518367) seeing in #626.

I've pushed two commits to remove the eager binding and then simplify the access patterns.

I'm including responses to your specific feedback below to elaborate on how this got tangled up, in case it's helpful for future work in this area:
___

> Here's what's bothering me...
>
> Including a DefaultSchema annotation on a _migration operation_ (which is what it means for to be in `NpgsqlMigrationsAnnotationProvider`) implies that there has been a change in the default schema, and which needs to be somehow applied to the database via SQL.

Right, that bothered me too at first, but I justified it with the assumption that we needed to allow for late-bound schema changes. But that's not right.

The mistake was binding one annotation to another in the first place. The default schema is describing to the _migration model_ where types should be placed if no schema is specified, so it shouldn't be consulted until the SQL is being generated for the migration. 

> Moreover, looking at the latest state of this PR, the DefaultSchema annotation doesn't seem to be used by NpgsqlMigrationsSqlGenerator... `GenerateEnumStatements()` iterates over all enum annotations on the AlterDatabaseOperation, which constructs PostgresEnum instances with fully-provided annotation name (already containing the schema) and value as strings.

That was the hiccup. Because of the way that the annotations were constructed with `schema.name` as the key, the annotation name may or may not have contained a schema. This was inconsistent when the source model had a `DefaultSchema` annotation, with the order of configuration playing into it.

The initial approach here was to "fix" this by doubling-down and yielding the `DefaultSchema` annotation to the `AlterDatabaseOperation`. But the answer is actually the exact opposite: _we should __never__ bind the value of one annotation into another_.

If no schema is given explicitly, then it's left null until generation-time when the default schema can be consulted from the appropriate model, rather than from the operation.

> In other words, unless I'm missing something, I think it's safer and more correct to consult the DefaultSchema only at creation time. If this is right, we need to fix this in range as well.

I think we agree here, assuming that when you say "at creation time", you mean when `Generate{Create,Drop}Enum` is called, rather than when the annotation is constructed.

roji replied:
I actually think we may be mis-communicating over this :)

> I think it stems from an early decision to have the DefaultSchema annotation eagerly bound into the enum annotation.

I think you actually had it right - see below.

> I think we agree here, assuming that when you say "at creation time", you mean when Generate{Create,Drop}Enum is called, rather than when the annotation is constructed.

By "creation time", I actually did mean annotation creation time, and not SQL generation time.

Let me explain. First, let's look at what happens when a migration is generated for regular table creation. For a model that has `.HasDefaultSchema("some_schema")`, EF Core will generate a create table migrations follows:

```c#
migrationBuilder.CreateTable(
    name: "some_table",
    schema: "some_schema",
    columns: ...);
```

In other words, the DefaultSchema is baked into the migration when it is generated. We don't get a migration with a null schema, deferring the "binding" with the default schema until SQL generation time - the schema is eagerly inserted into the generated migration. I think enum/range types should get the exact same treatment: the only difference is that unlike with tables, the schema (like everything else) is managed via annotations, but that shouldn't really make a difference.

Unless I'm mistaken, the idea here is that someone may add a migration creating a table, and then change the DefaultSchema on the model. If DefaultSchema is bound late, at SQL generation time, that means that the table is created with the new DefaultSchema, even though the DefaultSchema was modified after the table creation migration was created - probably not the desired behavior. I think the goal is that once a migration is created, it ideally has all the information needed to do everything, within itself, without any need to consult anything else. This doesn't work 100% in all scenarios, but we should definitely try to keep it that way as much as possible.

I admit it's a bit of a confusing situation, and I'm still not sure I have the mental model completely right about everything. If you still feel unsure about any of this we can reach out to one of the EF Core team members for verification.

austindrenski replied:
It sounds like we are on the same page for the design _goals_ (i.e. self-contained migrations). I'm just not sure how to _implement_ it with the current architecture...

> By "creation time", I actually did mean annotation creation time, and not SQL generation time.

> ```c#
> migrationBuilder.CreateTable(
>    name: "some_table",
>    schema: "some_schema",
>    columns: ...);
> ```
> In other words, the DefaultSchema is baked into the migration when it is generated. We don't get a migration with a null schema, deferring the "binding" with the default schema until SQL generation time - the schema is eagerly inserted into the generated migration. 

> I think enum/range types should get the exact same treatment: the only difference is that unlike with tables, the schema (like everything else) is managed via annotations, but that shouldn't really make a difference.

It really seems like management via annotations _does_ make a difference in this case.

The trouble I'm having with this is that when the default schema is eagerly bound into the annotation name, there's no recourse if a default schema is added (or updated) later in model configuration. For example:

```c#
NpgsqlConnection.GlobalTypeMapper.MapEnum<SomeEnum>("test.some_enum");

modelBuilder
    // The annotation name is constructed as `Npgsql:Enum:some_enum`
    .ForNpgsqlHasEnum<SomeEnum>()
    // But after this, it *ought* to be `Npgsql:Enum:test.some_enum`
    .HasDefaultSchema("test");

// 42704: type "test.some_enum" does not exist
```

In this case, late-binding solves the problem, albeit imperfectly.

_TBH I would rather see the default schema only ever used at generation-time so that I could move many tables/functions/types to a new schema by changing the default schema. But that's an opinion for another time..._

> I think the goal is that once a migration is created, it ideally has all the information needed to do everything, within itself, without any need to consult anything else. This doesn't work 100% in all scenarios, but we should definitely try to keep it that way as much as possible.

I agree that this should be our goal. 

Perhaps we're looking at this too late in the process. Would it be crazy to consider replace [`MigrationsModelDiffer`](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs)? 

It's currently in `Microsoft.EntityFrameworkCore.Migrations.Internal`, but it seems like it might give us the hooks to properly handle these as self-contained migration operations.

austindrenski replied:
> I think that rather than override `MigrationsModelDiffer`, this could be done cleanly in `NpgsqlMigrationsAnnotationProvider`, no? It's responsible for receiving annotations from the model and optionally providing annotations to the migrations process. That seems like a good spot to bake the default schema in, what do you think?

Oh, so rather than yield back the `DefaultSchema`, we would reconstruct the enum (and range) annotations and then yield back the "properly baked" version?

I hadn't considered that, but it sounds like it would work (and beats overriding a public-internal class). I'll give it a try this morning and report back.

roji replied:
@ajcvickers you may be interested in this conversation on default schema and when it's "bound" into other annotation which depend on it (e.g. annotation which defines an enum type on the model).

ajcvickers replied:
@AndriySvyryd @bricelam may have some thoughts on this.

austindrenski commented:
@roji I think this is very close to ready, but there are some changes since your last review that I'd like to get your OK on before merging. 

Specifically:

- `PostgresEnum.Labels` now has a private setter, which is used in place of calling `SetData` directly.
- I've removed the schema-less overload for `PostgresEnum.GetOrAddPostgresEnum(...)`. An equivalent form is still available via `model.Npgsql().GetOrAddPostgresEnum(...)`.
- I've removed `PostgresEnum.FindPostgresEnum`, which was only used within `PostgresEnum` and wasn't necessary. If this type of functionality is needed down the road, we could provide it via something like `model.Npgsql().FindPostgresEnum`.
- I've removed `PostgresEnum.BuildAnnotationName(...)`, per your review, as it was no longer necessary.
- I've updated `PostgresEnum.GetData()` to push the responsibility of returning a null-membered tuple down to `PostgresEnum.Deserialize(...)`. We were doing a null-check there anyways, so this just reduces some branching.

These changes are focused on simplifying `PostgresEnum` so that responsibility for convenience wrappers resides with classes like `NpgsqlModelAnnotations`.

_Also note I've split changes that were unrelated to the annotations into #724._

roji commented:
@austindrenski thanks for continuing to work on this. I'm not going to review again because of the DefaultSchema binding question - see my comment above. Once we resolve all that I'll give this another full review.

roji commented:
Thanks for continuing the work on this @austindrenski. I took a deeper dive into this and discovered some other problems, I'm sorry for not having done this before suggesting the MigrationsAnnotationsProvider approach. I know this is getting pretty long, but I think we'll get there soon.

First a more minor point. This PR indeed changes the binding of the default schema as discussed, but it also changes the API of PostgresEnum in general - the constructor now includes the previous "get or add" functionality, a new static `CreateAnnotation()` is an IAnnotation factory, `Labels` is no longer modifiable... When I first wrote PostgresEnum a while ago, I based it off the EF Core support for [Sequence](https://github.com/aspnet/EntityFrameworkCore/blob/6e69cad3903233f07e91ab717c195c9ee0644214/src/EFCore.Relational/Metadata/Internal/Sequence.cs), which is also an annotation managed on the model (after all, both are non-table, database-wide objects) - this is why the general API shape came from. As elsewhere, I believe there's a lot of value in keeping as aligned with how things work in EF Core, even if sometimes we think we can factor things better - it makes it very easy to switch between EF Core and EFCore.PG, and things work in a similar way. This is why I'm not too excited about redoing the general shape of PostgresEnum and friends.

This led me to reexamine how Sequence is implemented and interacts with migrations in general. Although at the metadata level it's just an annotation on the model (like our enum), with a Sequence object (similar to our PostgresEnum) to manage it, at the migration level it has special support in MigrationsModelDiffer - the latter detects it specifically much like it detects tables, and generates special `{Create,Drop}SequenceOperation` which the MigrationsSqlGenerator then renders into SQL. A long time ago, when I first started modelling PostgreSQL extensions (this was the first PostgreSQL-specific object I did), I actually started modifying the ModelDiffer (a bit like you suggested), but ran into some issues and the recommendation from the EF Core team was to avoid it - which is why everything is implemented as annotations instead, all the way into NpgsqlMigrationsSqlGenerator. For 3.0 we may want to revisit this, and reimplement extension/enum/range cleanly in the ModelDiffer. This would also move [the new/delete/alter detection](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/b56c93bcee2e3aec413b33a5a9b0fcee8bb9602d/src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs#L680) from the MigrationsSqlGenerator to the differ, where they conceptually belong. 3.0 would be the right milestone for this, as we can get any API required changes into ModelDiffer as necessary.

Even more important... When you change the DefaultSchema in EF Core, the ModelDiffer actually picks that up and generates Rename operations for *all* objects with an unspecified schema (so you get lots of `RenameTable()` and `RenameSequence()` calls in the new migration). This is implemented by having the annotation itself contain an empty/null schema, but having the [`Sequence.Schema` property](https://github.com/aspnet/EntityFrameworkCore/blob/6e69cad3903233f07e91ab717c195c9ee0644214/src/EFCore.Relational/Metadata/Internal/Sequence.cs#L128) return the default schema in that case (a bit like a previous version for this PR. So when the ModelDiffer compares the old list of sequences to the new, it detects the rename and generates a `RenameSequenceOperation`.

From a cursory look, I don't think it's possible for us to match this behavior with our annotation-only approach (i.e. without changing ModelDiffer). The ModelDiffer detects annotation changes, but the enum annotation itself hasn't changed at all - it still has a null schema; the separate DefaultSchema annotation is the one that changed.

Since we can't implement the standard EF Core behavior here (automatic rename on change of DefaultSchema), I suggest we not try to do anything fancy and leave things as simple as possible, and as close as possible to EF Core. This seems to be to do the same thing as Sequence - leave the schema null in the annotation, and return DefaultSchema from the schema property. This should create the annotation in the default schema initially, but not support automatically moving it if the default schema is changed (impossible to implement at the moment). We can definitely revisit this for 3.0.

Again, it's my fault for suggesting a bad path of work for you without looking deeper (sorry, many things going on at the moment...). I hope it's OK for you to revert this work and align it with EF Core as suggested above - or of course to propose any other approach for discussion.

PS IMHO we should maintain alignment with the EF Core codebase even when doing minor refactoring. For example, this PR moves PostgresEnum accessor properties to the top of the file, presumably because in some styles fields and properties come before constructors and methods. However, EF Core's [`Sequence.cs`](https://github.com/roji/EntityFramework/blob/6e69cad3903233f07e91ab717c195c9ee0644214/src/EFCore.Relational/Metadata/Internal/Sequence.cs) was written in a different style, which I think we should keep. I don't consider this terribly important and it's OK for us to deviate on this kind of thing - but I think it's important to always have the EF Core source code & design in front of you when making EFCore.PG changes, to be aware of exactly where we're deviating etc.

roji commented:
Adding to my already insanely-long message above, here's the old discussion on extending our own ModelDiffer and providing custom migration operations: https://github.com/aspnet/EntityFrameworkCore/issues/6524. This is useful because it discusses why we didn't end up going down that road - but it may be worth revisiting two years later.

roji commented:
Opened #730 to discuss the ModelDiffer/custom migration question again.

austindrenski commented:
> First a more minor point. This PR indeed changes the binding of the default schema as discussed, but it also changes the API of `PostgresEnum` in general - the constructor now includes the previous "get or add" functionality, a new static `CreateAnnotation()` is an `IAnnotation` factory, `Labels` is no longer modifiable... 

I recognize that these are departures from the previous design, and I can roll back some of it (e.g. file layout), but we do need _some_ of these changes. For example, `CreateAnnotation(...)` is what allows us to robustly return annotations in the annotation provider that are potentially different from the annotation that is actually serialized in the model.

Regarding the mutability of `Labels`, my primary concern is _what would actually be mutated_? Is it the model, a migration operation, a scaffolding model? The impact of setting that property depends on what the underlying annotation is, and if the underlying annotation is intended to be mutable. 

With those questions outstanding, and since we don't (yet!) support altering enums, I would rather keep these as immutable as possible, at least until we start to support alter operations on them.

> Since we can't implement the standard EF Core behavior here (automatic rename on change of `DefaultSchema`), I suggest we not try to do anything fancy and leave things as simple as possible, and as close as possible to EF Core. This seems to be to do the same thing as `Sequence` - leave the schema null in the annotation, and return `DefaultSchema` from the schema property.

That works for me. But I think that this does require something _like_ `CreateAnnotations(...)` to keep the annotation provider as clean as possible. 

I'm open to other ways to accomplish this, but I settled on the current design because it reduced the need to duplicate annotation logic and kept the "serialization" logic inside of `PostgresEnum`.

___ 

I've updated the PR to rollback some of the obvious stuff and rework the point of access for the default schema. Let me know how that looks, or if you think more needs to be rolled back (i.e. if settable properties are critical). 

I'll update the other PRs now, but don't worry about looking at them yet, I just want to make sure the patterns/logic look good in the other metadata classes too.

roji commented:
> I recognize that these are departures from the previous design, and I can roll back some of it (e.g. file layout), but we do need some of these changes. For example, CreateAnnotation(...) is what allows us to robustly return annotations in the annotation provider that are potentially different from the annotation that is actually serialized in the model.

My point was that this was not actually something we should be doing - I took back my idea of translating the model enum annotation into a different annotation for the migrations... The reasoning is that if the user doesn't specify a schema for the enum, then the generated migration should actually *not* contain the default schema (again, I reversed what I had previously written). Instead, the migration code should contain null for the schema - just like table and sequence do - and the binding should only occur very late, at the SQL migration phase. I recommend taking a look at [Sequence.cs](https://github.com/aspnet/EntityFrameworkCore/blob/6e69cad3903233f07e91ab717c195c9ee0644214/src/EFCore.Relational/Metadata/Internal/Sequence.cs#L19) for inspiration on how PostgresEnum.cs should look like, even if they're not going to be identical.

Just to be extra clear: my change of mind here really stemmed from a misunderstanding of how EF Core works... If a user doesn't specify a schema, the resulting migration should *not* contain a schema - even if the actual object gets creating in the default schema - and if the default schema later changes, then EF Core moves the object from the old default schema to the new one, at that very moment. Keeping the schema null is in fact required in order to be able to detect objects which were created in the (previous) default schema; if the schema were baked into the annotation, we wouldn't be able to know whether the user specified it (in which case it shouldn't move) or whether it was just the default (in which case it should).

> Regarding the mutability of Labels, my primary concern is what would actually be mutated? Is it the model, a migration operation, a scaffolding model? The impact of setting that property depends on what the underlying annotation is, and if the underlying annotation is intended to be mutable.

That's a valid question... I think that the mutability of the metadata (i.e. the annotations) isn't the same as our capability to carry out the corresponding change in the database. In other words, there may be scenarios where the metadata is simply built incrementally by the user, i.e. they interact directly with PostgresEnum to build the list of values based on some logic, rather than using the `ForNpgsqlHasEnum()` extension method. Don't forget that `PostgresEnum`, as well as `NpgsqlModelAnnotations` and the rest are completely public and accessible to the user. I think that's the point behind allowing mutation - just as a way of building the metadata on the model.

I'll really try to be available tonight (your morning) so we can have some online time together to discuss all this.

austindrenski commented:
OK, cleaned up and ready for review.

roji commented on src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs:
```diff
@@ -708,7 +709,7 @@ public virtual void Generate(NpgsqlDropDatabaseOperation operation, [CanBeNull]
             [NotNull] MigrationCommandListBuilder builder)
         {
-            var schema = GetSchemaOrDefault(enumType.Schema, model);
+            var schema = enumType.Schema ?? model.Relational().DefaultSchema;
```
I think that this logic should actually be inside `PostgresEnum.Schema`, follow the [sequence example](https://github.com/aspnet/EntityFrameworkCore/blob/6e69cad3903233f07e91ab717c195c9ee0644214/src/EFCore.Relational/Metadata/Internal/Sequence.cs#L128) as much as possible.

austindrenski replied:
I would prefer that too, but the `PostgresEnum._annotatable` will not always have the default schema set, even where it is logically "active".

That said, this logic is actually present in the getter for `PostgresEnum.Schema`:

```c#
public string Schema => Deserialize(Annotation).Schema ?? (string)_annotatable[RelationalAnnotationNames.DefaultSchema];
```

However, that logic is insufficient for migrations since we forward the enum annotation to `AlterDatabaseOperation`, but [we do not want to forward](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/685#discussion_r236064018) the `DefaultSchema` annotation.

So in this specific case, `enumType._annotatable` is the `AlterDatabaseOperation`, not the `model`. So calling `enumType.Schema` would return `null` even with the `Sequence`-like getter logic.

roji replied:
Ok, will do my best to take another last look in the next few hours so we can lock this down...

roji replied:
OK, I see your point. You're right that this logic cannot work inside the annotation.

In that case, should we not be removing the null-coalescing logic from the annotation, and simply return null from its Schema property? It feels wrong to have both.

austindrenski replied:
Good point, plus it's probably safer to remove it so there's no ambiguity over whether the schema is set or not.

roji commented on src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs:
```diff
@@ -735,10 +736,10 @@ public virtual void Generate(NpgsqlDropDatabaseOperation operation, [CanBeNull]
             [NotNull] MigrationCommandListBuilder builder)
         {
-            var schema = GetSchemaOrDefault(enumType.Schema, oldDatabase);
+            var schema = enumType.Schema ?? model.Relational().DefaultSchema;
```
As above

roji commented on src/EFCore.PG/Metadata/PostgresEnum.cs:
```diff
@@ -9,102 +9,126 @@
+                    .Select(a => new PostgresEnum(annotatable, a.Name));
+
+        [CanBeNull]
+        IAnnotation Annotation => _annotatable.FindAnnotation(_annotationName);
```
I'd suggest reverting this, aligning ourselves with the EF Core practice here. Properties such as `Name` can simply call `GetData()` as before, obviating the need for a private `Annotation` property. It's about keeping the code as similar as possible to how things are done in EF Core, unless there's a compelling reason to do otherwise.

austindrenski replied:
As below, this mattered more in a previous iteration of this PR. Will revert.

roji commented on src/EFCore.PG/Metadata/PostgresEnum.cs:
```diff
@@ -9,102 +9,126 @@
-        public static IEnumerable<PostgresEnum> GetPostgresEnums([NotNull] IAnnotatable annotatable)
         {
             Check.NotNull(annotatable, nameof(annotatable));
+            Check.NullButNotEmpty(schema, nameof(schema));
```
I'd prefer reverting most (all?) of the changes here to `GetOrAddPostgresEnum()` and bring back `FindPostgresEnum()`. Even if the latter isn't currently used in our code base, it's a basic metadata lookup tool, which is once again present in EF Core's sequence implementation (and other similar metadata objects).

austindrenski replied:
The API changes mattered more for earlier iterations of the PR, when some of the logic was still fuzzy and incomplete. But at this point, I think we have the semantics clear enough that the public API changes don't have much value anymore.

I'll revert most of this now.

roji replied:
Makes sense...

austindrenski commented:
OK, the only notable divergence in the public API is the removal of the schema-free overload to `GetOrAddPostgresEnum`. 

Personally, I don't see the value in that overload since the annotation classes provide user-friendly overloads, but I can revert to include it if you think it still has some value.

austindrenski commented on src/EFCore.PG/Metadata/PostgresEnum.cs:
```diff
@@ -9,102 +9,163 @@
-
+            => schema != null
+                ? $"{NpgsqlAnnotationNames.EnumPrefix}{schema}.{name}"
+                : $"{NpgsqlAnnotationNames.EnumPrefix}{name}";
```
Good idea. Will have to punt for 2.2.1 though.

roji replied:
Hey @AndriySvyryd :) Didn't see your comment before submitting my review below...

* Any specific reason for to have a single annotation for all enums?
* The current scheme is already in place, so unless I'm mistaken a change here would break existing migration code...
* Is annotation lookup/traversal considered a perf-sensitive area? I was under the impression it's only used at initialization/migration generation time...

roji replied:
> I haven't checked how this is used, but potentially the annotation could be looked up every time a value of this type is sent to the database.

I'm not sure how this could be happening (unless I'm forgetting something)... At least for this specific annotation - define an enum type - this is purely an annotation on the model itself, which should only be picked up at the migrations SQL generation level for creating the appropriate PostgreSQL enum type... The other part of the enum support is at the TypeMappingSource, but that happens as a context option and isn't related to the model or to any annotations.

But we should definitely be sure of the perf situation here...

Regardless, as @austindrenski said, it's a bit late to do this for 2.2.0 (we're planning to release in a few hours :)). But still definitely important...

AndriySvyryd replied:
If it's just migrations (like `Sequence`) then you are correct that it doesn't matter much, but it might be worthwhile to check where this and other annotations that have constructed names are used.

roji replied:
OK, thanks for the tip - we'll definitely take a look at this after the release to make sure.

austindrenski replied:
Opened #736 to track further discussion after `2.2.0`.

austindrenski commented on src/EFCore.PG/Metadata/PostgresEnum.cs:
```diff
@@ -9,102 +9,163 @@
-        }
+            var annotationName = BuildAnnotationName(schema, name);
 
-        public static PostgresEnum GetOrAddPostgresEnum(
```
Okay, will revert.

austindrenski commented:
@roji Just updated. Are you able to initiate the release of `4.0.4` (driver + plugins)?
